flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/RowToRowCastRule.java

```diff
@@ -221,7 +221,8 @@ protected String generateCodeBlockInternal(
                                     elseBodyWriter.stmt(writeNull));
         }
 
-        writer.stmt(methodCall(writerTerm, "complete")).assignStmt(returnVariable, rowTerm);
+        writer.stmt(methodCall(writerTerm, "complete"))
+                .assignStmt(returnVariable, rowTerm + ".copy()");
```

Comment:
BTW, the current design is not very good. It should let the outter to determine whether we should copy or not. For this case, it's better that ArrayToArrayCastRule to determine whether the inner block should reuse or not if the object is reused.